src/rgw/rgw_rados.cc

```diff
@@ -8163,11 +8191,24 @@ int RGWRados::bucket_rebuild_index(RGWBucketInfo& bucket_info)
 {
   librados::IoCtx index_ctx;
   map<int, string> bucket_objs;
-  int r = open_bucket_index(bucket_info, index_ctx, bucket_objs);
-  if (r < 0)
+
+  /* handle on going bucket resharding */
+  int r = reshard->block_while_resharding(bucket_info.bucket.oid);
+  if (r < 0) {
     return r;
+  }
 
-  return CLSRGWIssueBucketRebuild(index_ctx, bucket_objs, cct->_conf->rgw_bucket_index_max_aio)();
+  r = open_bucket_index(bucket_info, index_ctx, bucket_objs);
+  if (r < 0) {
+    reshard->unlock_bucket_index(bucket_info.bucket.oid);
```

Comment:
@yehudasa , we have to lock so it won't change during the rebuild.
I will add a Guard calls for the lock , that will free it atomaticly